Fix an incorrect CSE case with struct retyping.

sandreenko opened the pull request:
The failing assert was:
```
Assertion failed 'IsCompatibleType(cseLclVarTyp, expTyp)'  
F:\git\runtime\src\coreclr\src\jit\optcse.cpp Line: 2768
```

That was a combination of two issues:
1.  There was an `IND struct` under `ADDR` not marked with `DONT_CSE`, it came from `INDEX` morphing;
2. There was another `IND` for the same `LCL_VAR`, but as `ref` from call return retyping, `VNApplySelectorsTypeCheck` saw this `IND ref` and returned `struct` type for it without checking sizes/structHandles (we don't have handles in this case, because we don't have a mechanism to keep them on pointers and we access array through a ref (pointer) local var);
-> CSE saw `IND struct` and `IND ref` with the same VN and without `DONT_CSE` and failed.

The first issue was fixed the third commit https://github.com/dotnet/runtime/pull/34676/commits/86a9f22f9aa284404e32e7699c8042a9a78c5be1, 
it introduced a few diffs (20 methods in frameworks, then a few in BING/pri1 SPMI), all textual, not size diffs. I have fixed them in https://github.com/dotnet/runtime/pull/34676/commits/cd24383210b9ed22f7ea3640005217021f9fbd63 commit.

The solution adds this flag on each node under `ADDR` and checks it in `debugCheckFlags`. It is questionable because we will for sure forget to clear `DONT_CSE` flag from time to time and we don't have a checker to assert when it is set without a reason, as we do for `CALL, ASG, EXCEPT` flags. 
We could add a checker (in another PR) but it won't be cheap because:
1. `DONT_CSE` is used not only for correctness but for profitability as well. I had a change that cleaned it from `NULLCHECK` nodes, but diffs were negative: CSE started to create CSE copies for them, but the null checks were later deleted by assertion propagation, so changing CSE logic to have zero diffs would be non-trivial;
2. we don't have a clear contract where `DONT_CSE` is required for the correctness and there are comments with TODO to delete some that reference failing tests if you do.

Maybe it would be better to drop the commit for SPMI diffs and accept textual diffs instead of adding these non-obvious conditions. I am sure there are some other places where we forgot to clean that, they just don't produce any diffs.

Commits:
61286138a07: Fix old printing issues.
df5266c8285: add a repro test that doesn't require crossgen2.
86a9f22f9aa: Check that all `ADDR` sources are marked as DONT_CSE.
cd24383210b: Fix text(no size) diffs found by SPMI (could be dropped).
c0300c46777: Add a question/comment (will be dropped/fixed before merge).

Fixes #33884.

Commit:
Fix old printing issues.

Commit:
add a repro test that doesn't require crossgen2.

Commit:
Check that all `ADDR` sources are marked as DONT_CSE.

Commit:
Fix text(no size) diffs found by SPMI.

Commit:
Add a question/comment.

sandreenko commented:
PTAL @erozenfeld @briansull, cc @dotnet/jit-contrib 
I would like to hear your opinion about adding a checker for `DONT_CSE` or move its setting to another place/phase.

briansull commented on src/coreclr/src/jit/morph.cpp:
```diff
@@ -16831,8 +16833,15 @@ void Compiler::fgMorphStructField(GenTree* tree, GenTree* parent)
+                bool canCSE  = tree->CanCSE();
+                // That doesn't make any sense, was it supposed to clean all flags except GTF_NODE_MASK here? 2012<
                 tree->gtFlags &= GTF_NODE_MASK;
+                // If yes then why a year after (2013) was this cleaning added, that obviously does nothing.
```
I don't think that you want to checkin these two comments.

sandreenko replied:
Yes, I will drop that commit before merge, want to see if somebody knows more about these changes.

briansull commented on src/coreclr/src/jit/importer.cpp:
```diff
@@ -9121,7 +9123,12 @@ GenTree* Compiler::impFixupStructReturnType(GenTree* op, CORINFO_CLASS_HANDLE re
+            op          = op1->AsOp()->gtOp1;
+            if (canCSE)
+            {
+                op->ClearDoNotCSE();
```
It isn't necessary to guard this call to ClearDoNotCSE()

sandreenko replied:
Why is not it? It could be a volatile LCL_VAR that needs that flag.

briansull replied:
The implementation of CanCSE is 
` return ((gtFlags & GTF_DONT_CSE) == 0);`
So this is basically saying if the flag is set then clear it using the AND NOT operation.
Not sure how a volatile LCL_VAR would behave any differently here

sandreenko replied:
> So this is basically saying if the flag is set then clear it using the AND NOT operation.

Maybe I am missing something, but I don't understand.
We are checking the flag on `IND(ADDR(X))` tree and clear it on `X`. `X` was marked as `DONT_CSE` because it was `ADDR` source, `IND(ADDR(X))` could be marked as `DONT_CSE` for other reasons that we don't change here, so if `IND(ADDR(X))` was marked with that flag then after the transformation `X` should have this flag as well.

For example, `IND1(ADDR2(IND3(ADDR4(X))` in the beginning `X` and `IND3` have `DONT_CSE` because they are `ADDR` sources, when we optimize `IND3(ADDR4(X))` we have to keep `DONT_CSE` on `X` .

briansull replied:
Yes, I see.
I missed that we were overwriting `op` after setting `canCSE`.
This sequence is confusing and could use a comment, saying we are preserving any DONT_CSE that was set on the original version of `op`

```
 bool canCSE = op->CanCSE();
          op        = op1->AsOp()->gtOp1;
```

briansull commented on src/coreclr/src/jit/importer.cpp:
```diff
@@ -3863,11 +3863,13 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
-            GenTree*             byReferenceStruct = gtCloneExpr(thisptr->gtGetOp1());
-            assert(byReferenceStruct != nullptr);
+            GenTreeLclVar*       byReferenceStruct = gtCloneExpr(lclVar)->AsLclVar();
+            assert(!byReferenceStruct->CanCSE());
```
Did you want to remove the assert?  I believe that gtCloneExpr can return nullptr for some complex trees.
`assert(byReferenceStruct != nullptr);`

sandreenko replied:
`CanCSE` does `this` dereferencing, so it is not necessary to check it separately, but I will return it if it reads better.

briansull replied:
An A/V in a Checked build is worse than an assert firing, in terms of debuggablity.

briansull commented:
> I would like to hear your opinion about adding a checker for DONT_CSE or move its setting to another place/phase.

Checker phases are typically used to enforce correctness issues.
The DONT_CSE is a flag that prevents an optimization, presumably because it is unsafe.
When code sets it is could be unsafe for all cases or if is is difficult to determine it could be unsafe just for some cases.

Failing to perform a CSE isn't a correctness issue and often won't even significantly change the performance of the generated code.   It will show up in code diffs, which looks like what your bug fix is trying to correct here.

> Maybe it would be better to drop the commit for SPMI diffs and accept textual diffs instead of adding these non-obvious conditions. 

I am fine with accepting a small amount of textual diffs and we often get them when code changes the behavior of the CSE heuristics .

sandreenko commented:
> Checker phases are typically used to enforce correctness issues.
> The DONT_CSE is a flag that prevents an optimization, presumably because it is unsafe.
> When code sets it is could be unsafe for all cases or if is is difficult to determine it could be unsafe just for some cases.

How is it different from, for example, `GTF_EXCEPT`? When we don't set `GTF_EXCEPT` where we need it - it is a correctness issue. When we set it if we don't need it - it forbids some optimizations.

briansull commented:
GTF_EXCEPT flags are a bit different IMO,  these are flags that get push up the tree to all of the parent nodes.
Thus it summarizes the state below, so that we don't have to walk the entire tree to look for operations that have exceptional side effects.   Since when inserting new node it is easy to forget to set the GTF_EXCDEPT flags , a checker is great to insure that the summery is always correct.  I am sure that correctness of our optimization depend upon having correct GTF_EXCEPT flags, although having an extra flag would only disable an optimization (similar to having an extra DONT_CSE flag)

The DONT_CSE flag is a local flag that only applies to the current node, and typically the parent node is responsible for setting it on one or both of its child nodes.

That said I am not strongly opposed to adding some kind of checker if it provides good value.

erozenfeld commented:
We currently have 56 places in the code where we set `GTF_DONT_CSE` and an additional 11 places where we propagate it. I think it would be beneficial to audit these places and do three things:
1. For places where there are comments that indicate that it's not clear why the flag is set, try not to set it and figure out the right solution.
2. For places that where there is a good reason to set the flag for correctness, see if it's possible to add a check to the checker to make sure we don't lose the flag.
3. For places where we set the flag for performance reasons we probably don't want to add a check to the checker but we should make sure there is a good comment explaining why the flag is set.
That will also make it impossible to add a check in the checker that we shouldn't see an unexpected `GTF_DONT_CSE`.

sandreenko commented on src/coreclr/src/jit/valuenum.cpp:
```diff
@@ -3614,6 +3614,13 @@ ValueNum ValueNumStore::VNApplySelectorsTypeCheck(ValueNum elem, var_types indTy
 
         if ((indType == TYP_REF) && (varTypeIsStruct(elemTyp)))
         {
+            // This whole block is over-optimistic, we don't have any information
```
I checked that if we delete that unsafe condition we will get 8 bytes regression in framework assemblies (1 method) and 2 in SPMI (pri1 and BING).
That regression will be fixed with #33225, so I think I will just delete this block before the merge.

erozenfeld commented on src/coreclr/src/jit/morph.cpp:
```diff
@@ -5376,7 +5376,12 @@ GenTree* Compiler::fgMorphArrayIndex(GenTree* tree)
+        if (!canCSE)
+        {
+            indir->SetDoNotCSE();
+        }
```
Why do you need to check whether`DONT_CSE` is set on `indir` here before setting it? Why not just set it?

erozenfeld replied:
Never mind, I misread the change.

erozenfeld commented on src/coreclr/src/jit/morph.cpp:
```diff
@@ -16831,8 +16833,15 @@ void Compiler::fgMorphStructField(GenTree* tree, GenTree* parent)
+                if (!canCSE)
+                {
+                    tree->SetDoNotCSE();
+                }
```
Same comment: why not set `DONT_CSE` without checking it first?

erozenfeld replied:
Never mind, I misread the change.

erozenfeld commented on src/coreclr/src/jit/morph.cpp:
```diff
@@ -16831,8 +16833,15 @@ void Compiler::fgMorphStructField(GenTree* tree, GenTree* parent)
                 tree->AsLclVarCommon()->SetLclNum(fieldLclIndex);
                 tree->gtType = fieldType;
+                bool canCSE  = tree->CanCSE();
+                // That doesn't make any sense, was it supposed to clean all flags except GTF_NODE_MASK here? 2012<
```
I can help you find the change that introduced this.

erozenfeld reviewed (commented):
Left a few comments.

Commit:
Update src/coreclr/src/jit/compiler.hpp

Co-Authored-By: Eugene Rozenfeld <erozen@microsoft.com>

Commit:
Delete an optimization that was not proved to be correct.

Commit:
Delete useless code.

Commit:
Revert "Fix text(no size) diffs found by SPMI."

This reverts commit cd24383210b9ed22f7ea3640005217021f9fbd63.

sandreenko commented on src/coreclr/src/jit/valuenum.cpp:
```diff
@@ -3612,15 +3612,7 @@ ValueNum ValueNumStore::VNApplySelectorsTypeCheck(ValueNum elem, var_types indTy
         size_t elemTypSize = (elemTyp == TYP_STRUCT) ? elemStructSize : genTypeSize(elemTyp);
         size_t indTypeSize = genTypeSize(indType);
 
-        if ((indType == TYP_REF) && (varTypeIsStruct(elemTyp)))
```
That change gives as one regression in framework crossgening:
```
8 (12.70% of base) : Newtonsoft.Json.dasm -  Newtonsoft.Json.Serialization.JsonTypeReflector:get_DynamicCodeGeneration():bool
```

sandreenko commented:
I have updated the PR, deleted the suspicious optimization from `VNApplySelectorsTypeCheck` and allowed textual diffs where we forget to clean DONT_CSE.
Please take another look.

briansull reviewed (approved):
LGTM now

erozenfeld reviewed (approved):
LGTM
